### PR TITLE
Fix Log Replication bug in backup restore workflow

### PR DESCRIPTION
**Pull request by zhangn49**
## Overview

Description:

In backup restore workflow, the active cluster can be replaced by a backup cluster, whose corfu log tail might be regressed. 
The standby cluster will ignore all updates until the backup cluster's log address go beyond its snapshot mark.

So the logEntryWriter's snapshot mark should be reset when there is a snapshot sync just completes. In this way, log entry sync will properly replicate data.

Why should this be merged: 

Bug Fix

Related issue(s) (if applicable): #2980 

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

**Comment by zhangn49** [comment]
Hi @annym The base branch is your remove_getTimestamp branch, as it carries multiple LR changes.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogEntryWriter.java [hunk 1]
```diff
@@ -155,6 +146,8 @@ public long apply(LogReplicationEntryMsg msg) throws ReplicationWriterException
 
         // A new Delta sync is triggered, setup the new srcGlobalSnapshot and msgQ
         if (msg.getMetadata().getSnapshotTimestamp() > srcGlobalSnapshot) {
+            log.warn("setup the new srcGlobalSnapshot and lastMsgTs {}",
```
Comment (annym) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogEntryWriter.java:149]: Can this message be something a bit more significant or easy to follow in logs... Perhaps something like: "New log entry sync has started. Higher snapshot detected. prev={}, new={}" or something along those lines.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogEntryWriter.java [hunk 2]
```diff
@@ -183,7 +176,7 @@ public long apply(LogReplicationEntryMsg msg) throws ReplicationWriterException
      * that is the last log entry it has played.
      * This is called while the writer enter the log entry sync state.
      *
-     * @param snapshot
+     * @param snapshot the base snapshot that last full sync based on.
```
Comment (annym) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogEntryWriter.java:179]: nit -> the base snapshot on which the last full sync was based on.

File: test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java [hunk 3]
```diff
@@ -1007,6 +1011,140 @@ public void testActiveLockRelease() throws Exception {
         assertThat(mapStandby.size()).isEqualTo(secondBatch);
     }
 
+    /**
+     * This test verifies log entry works after a force snap sync
```
Comment (annym) [inline test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:1015]: nit -> snap = snapshot

File: test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java [hunk 4]
```diff
@@ -1007,6 +1011,140 @@ public void testActiveLockRelease() throws Exception {
...
+     * in the backup/restore workflow.
+     * <p>
+     * 1. Init with corfu 9000 active, 9001 standby, 9002 backup
+     * 2. Write 10 entries to active map
+     * 3. Start log replication: Node 9010 - active, Node 9020 - standby, Node 9030 - backup
+     * 4. Wait for Snapshot Sync, both maps have size 10
+     * 5. Write 5 more entries to active map, to verify Log Entry Sync
+     * 6. Change the topology, backup 9030 will be the new active
```
Comment (annym) [inline test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:1023]: I think these descriptions are nice, but here we are missing some steps as it is not clear that backup lags behind, maybe we can add this info so it fully explains the test.

File: test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java [hunk 5]
```diff
@@ -1007,6 +1011,140 @@ public void testActiveLockRelease() throws Exception {
...
+        CorfuTable<String, Integer> mapBackup = backupRuntime.getObjectsView()
+                .build()
+                .setStreamName(streamName)
+                .setTypeToken(new TypeToken<CorfuTable<String, Integer>>() {
+                })
+                .open();
+
+        assertThat(mapBackup.size()).isZero();
+
+        // Write 10 entries to active map
```
Comment (annym) [inline test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:1049]: This is nice, can you add a comment. Like "Initializing corfu servers with dummy data, to confirm data is wiped on startup" cause that is the purpose right?

File: test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java [hunk 6]
```diff
@@ -1007,6 +1011,140 @@ public void testActiveLockRelease() throws Exception {
...
+        assertThat(mapBackup.size()).isEqualTo(secondBatch);
+
+        // Wait until data is fully replicated again
+        waitForReplication(size -> size == secondBatch, mapStandby, secondBatch);
+        log.info("After delta sync, both maps have size {}. Current backup corfu[{}] log tail " +
+                        "is {}, standby corfu[{}] log tail is {}", firstBatch, backupClusterCorfuPort,
+                backupRuntime.getAddressSpaceView().getLogTail(), standbyClusterCorfuPort,
+                standbyRuntime.getAddressSpaceView().getLogTail());
+
+        shutdownCorfuServer(backupCorfu);
```
Comment (annym) [inline test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:1144]: Should you shutdown active and standby as well?

File: test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java [hunk 7]
```diff
@@ -1007,6 +1011,140 @@ public void testActiveLockRelease() throws Exception {
...
+        shutdownCorfuServer(backupReplicationServer);
+    }
```
Comment (annym) [inline test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:1146]: Nice Test!

**Review by annym — APPROVED** [review]

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogEntryWriter.java (hunk 2 above)
Comment (zhangn49) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogEntryWriter.java]: done.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogEntryWriter.java (hunk 1 above)
Comment (zhangn49) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogEntryWriter.java]: done.

File: test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java (hunk 3 above)
Comment (zhangn49) [reply test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java]: done

File: test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java (hunk 4 above)
Comment (zhangn49) [reply test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java]: yes, added a few more description about the backup

File: test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java (hunk 5 above)
Comment (zhangn49) [reply test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:1051]: You mean dummy data on the standby side, right? at line 1059.
And it is used to make sure the standby corfu log tail is much longer than the backup one.

File: test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java (hunk 6 above)
Comment (zhangn49) [reply test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:1148]: Yes, they are shutdown by the @After function

**Review by pankti-m — APPROVED** [review]

**Pull request merged**